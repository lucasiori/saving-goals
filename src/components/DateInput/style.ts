import styled from 'styled-components';

const Wrapper = styled.div`
  position: relative;
  min-width: 215px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  border: ${({ theme }) =>
    `${theme.borders.width.sm} solid ${theme.colors.blueGray['50']}`};
  border-radius: ${({ theme }) => theme.borders.radius.sm};
  padding: ${({ theme }) => `${theme.spacings.xxxs} ${theme.spacings.xs}`};

  &[data-focused='true'] {
    border: ${({ theme }) =>
      `${theme.borders.width.lg} solid ${theme.colors.dark}`};
  }

  input {
    position: absolute;
    width: ${({ theme }) => `calc(100% - (${theme.spacings.xs} * 2))`};
    opacity: 0;
    z-index: -1;
  }
`;

const ArrowControl = styled.button`
  width: 24px;
  height: 24px;
  display: flex;
  align-items: center;
  justify-content: center;
  background: none;
  border: 0;

  &[disabled] {
    cursor: not-allowed;
    opacity: 0.5;
  }
`;

const Content = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin: ${({ theme }) => `0 ${theme.spacings.md}`};

  strong,
  span {
    font-size: ${({ theme }) => theme.font.sizes.xs};
    line-height: ${({ theme }) => theme.font.lineHeights.lg};
  }

  strong {
    color: ${({ theme }) => theme.colors.blueGray['900']};
  }

  span {
    color: ${({ theme }) => theme.colors.blueGray['400']};
  }
`;

export { Wrapper, ArrowControl, Content };
