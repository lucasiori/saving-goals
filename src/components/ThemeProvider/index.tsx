import { ThemeProvider } from 'styled-components';
import theme from '../../styles/theme';

interface ThemeProviderProps {
  children: React.ReactNode;
}

export default function ThemeProviderCP({
  children,
}: ThemeProviderProps): JSX.Element {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}
