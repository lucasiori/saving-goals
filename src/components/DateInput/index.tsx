import { useEffect, useState } from 'react';
import { format } from 'date-fns';
import useDate from '../../hooks/use-date';
import chevronLeftSvg from '../../assets/icons/chevron-left.svg';
import chevronRightSvg from '../../assets/icons/chevron-right.svg';
import { Wrapper, ArrowControl, Content } from './style';

interface DateInputProps {
  id: string;
  onChangeDate?: (month: number, year: number) => void;
}

export default function DateInputCP({
  id,
  onChangeDate = () => {},
}: DateInputProps): JSX.Element {
  const {
    selectedMonth,
    selectedYear,
    hasPreviousMonth,
    prevMonth,
    nextMonth,
  } = useDate();

  const [isFocused, setIsFocused] = useState(false);

  function handleKeyUp({ key }: React.KeyboardEvent<HTMLDivElement>) {
    if (key === 'ArrowLeft' && hasPreviousMonth) {
      prevMonth();
    } else if (key === 'ArrowRight') {
      nextMonth();
    }
  }

  useEffect(() => {
    onChangeDate(selectedMonth, selectedYear);
  }, [selectedMonth, selectedYear, onChangeDate]);

  return (
    <Wrapper data-focused={isFocused}>
      <input
        data-id="date-input"
        type="text"
        id={id}
        name={id}
        onFocus={() => setIsFocused(true)}
        onBlur={() => setIsFocused(false)}
        onKeyUp={handleKeyUp}
        onChange={() => {}}
        value={`${selectedYear}-${selectedMonth}`}
      />

      <ArrowControl
        data-id="prev-button"
        type="button"
        onClick={prevMonth}
        disabled={!hasPreviousMonth}
      >
        <img src={chevronLeftSvg} alt="Previous" width="24px" />
      </ArrowControl>

      <Content>
        <strong>{format(new Date().setMonth(selectedMonth), 'MMMM')}</strong>
        <span>{format(new Date().setFullYear(selectedYear), 'yyyy')}</span>
      </Content>

      <ArrowControl data-id="next-button" type="button" onClick={nextMonth}>
        <img src={chevronRightSvg} alt="Next" width="24px" />
      </ArrowControl>
    </Wrapper>
  );
}
