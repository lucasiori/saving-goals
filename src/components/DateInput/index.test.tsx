import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import ThemeProvider from '../ThemeProvider';
import DateInput from '.';

describe('DateInput', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <DateInput id="date-input-test" />
      </ThemeProvider>
    );
    const component = tree.toJSON();

    expect(component).toMatchSnapshot();
  });

  describe('when there are not previous month', () => {
    it('should disabled previous button', () => {
      const component = mount(
        <ThemeProvider>
          <DateInput id="date-input-test" />
        </ThemeProvider>
      );

      expect(
        component.find('button[data-id="prev-button"]').prop('disabled')
      ).toBeTruthy();
    });
  });

  describe('when click on control buttons', () => {
    it('should trigger onChangeDate function', () => {
      const onChangeDate = jest.fn();

      const component = mount(
        <ThemeProvider>
          <DateInput id="date-input-test" onChangeDate={onChangeDate} />
        </ThemeProvider>
      );

      component.find('button[data-id="next-button"]').simulate('click');
      component.find('button[data-id="prev-button"]').simulate('click');

      // 1° / 2º call - render
      // 3° call - next
      // 4° call - prev

      expect(onChangeDate).toBeCalledTimes(4);
    });
  });

  describe('when click on left and right arrows', () => {
    it('should trigger onChangeDate function', () => {
      const onChangeDate = jest.fn();

      const component = mount(
        <ThemeProvider>
          <DateInput id="date-input-test" onChangeDate={onChangeDate} />
        </ThemeProvider>
      );

      component.find('input[data-id="date-input"]').simulate('focus');
      component
        .find('input[data-id="date-input"]')
        .simulate('keyup', { key: 'ArrowRight' });
      component
        .find('input[data-id="date-input"]')
        .simulate('keyup', { key: 'ArrowLeft' });

      // 1° / 2º call - render
      // 3° call - next
      // 4° call - prev

      expect(onChangeDate).toBeCalledTimes(4);
    });
  });
});
