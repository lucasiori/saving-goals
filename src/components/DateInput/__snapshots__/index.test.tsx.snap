// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput renders correctly 1`] = `
<div
  className="sc-bdnxRM gQFGBg"
  data-focused={false}
>
  <input
    data-id="date-input"
    id="date-input-test"
    name="date-input-test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    type="text"
    value="0-0"
  />
  <button
    className="sc-gtsrHT iuILPr"
    data-id="prev-button"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <img
      alt="Previous"
      src="chevron-left.svg"
      width="24px"
    />
  </button>
  <div
    className="sc-dlnjwi jFDLXd"
  >
    <strong>
      January
    </strong>
    <span>
      0001
    </span>
  </div>
  <button
    className="sc-gtsrHT iuILPr"
    data-id="next-button"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="Next"
      src="chevron-right.svg"
      width="24px"
    />
  </button>
</div>
`;
